### PR TITLE
Analytics improvements for future use.

### PR DESCRIPTION
**Pull request by rkuykendall**

### CODE REVIEW

**Comment by diezguerra** [comment]
@rkuykendall what up with this?

**Comment by rkuykendall** [comment]
@diezguerra Wrote it but didn't have time to actually test and deploy. It's my pattern for how we should be doing analytics in the future. I think this is still a slower week on team withdraws so I'll try to verify and release it.

**Comment by diezguerra** [comment]
Cool. Grab a copy of that product workflow book (there's many in the office), it should give you a good idea of why letting things sit is bad.

File: src/analytics/TrackedLink.jsx [hunk 1]
```diff
@@ -35,8 +35,11 @@ class TrackedLink extends React.Component {
 
     _handleClick(event) {
         let data = omit(this.props, 'children', 'className', 'href', 'target');
+        data = {url: this.props.href, ...data};
+        let event = `clicked-${global.__env.config.appDisplayName}-${this.props.type}`;
```
Comment (learykara) [inline src/analytics/TrackedLink.jsx]: `config` should come through props-- avoid using `global` wherever possible

File: src/analytics/TrackedLink.jsx [hunk 2]
```diff
@@ -35,8 +35,11 @@ class TrackedLink extends React.Component {
 
     _handleClick(event) {
         let data = omit(this.props, 'children', 'className', 'href', 'target');
+        data = {url: this.props.href, ...data};
```
Comment (learykara) [inline src/analytics/TrackedLink.jsx]: hmm - you're omitting `href` from `this.props` in creating `data`, but then adding it back in with the name `url`?

File: src/analytics/TrackedLink.jsx (hunk 1 above)
Comment (learykara) [inline src/analytics/TrackedLink.jsx]: also any reason `event` shouldn't be a `const`?

File: src/analytics/TrackedLink.jsx [hunk 3]
```diff
@@ -35,8 +35,11 @@ class TrackedLink extends React.Component {
...
+        let event = `clicked-${global.__env.config.appDisplayName}-${this.props.type}`;
+        log(event, data);
```
Comment (learykara) [inline src/analytics/TrackedLink.jsx]: is this `log` in here intentional?

File: src/analytics/actions.es6 [hunk 4]
```diff
@@ -1,20 +1,57 @@
...
+    // Temporary stub, which can be expanded upon to add data.
+    global.analytics &&
+        global.analytics.page.apply(this, arguments);
+}
+
 module.exports = {
-    track
+    track: track,
+    identify: identify,
+    page: page
```
Comment (learykara) [inline src/analytics/actions.es6]: can this just be
```
module.exports = {
    track,
    identify,
    page
}
```
?

File: src/analytics/actions.es6 [hunk 5]
```diff
@@ -1,20 +1,57 @@
...
+// This event mirrors the call signature of global.analytics.track
+function track(event, properties, options, fn) {
+    // Argument reshuffling, from original library.
+    if (is.fn(options)) fn = options, options = null;
+    if (is.fn(properties)) fn = properties, options = null, properties = null;
+
+    appInfo = {
         app: result(global.__env.config, 'app.name', '').toLowerCase(),
         appDisplayName: result(global.__env.config, 'app.displayName', '').toLowerCase(),
         ...data
```
Comment (learykara) [inline src/analytics/actions.es6]: where does `data` come from in this fn?

File: src/analytics/actions.es6 [hunk 6]
```diff
@@ -1,20 +1,57 @@
...
     }
+    let __env = __env || {};
```
Comment (learykara) [inline src/analytics/actions.es6]: does this pick up `global.__env`?

File: src/analytics/TrackedLink.jsx (hunk 1 above)
Comment (learykara) [inline src/analytics/TrackedLink.jsx]: oh wait, it's an arg. this throws an error

File: src/analytics/actions.es6 [hunk 7]
```diff
@@ -1,20 +1,57 @@
 const log = require('debug')('ui:analytics');
 const result = require('lodash/object/result');
+const Qs = require('qs');
 
-function track(action, type, data={}) {
-    action = `${action}: ${global.__env.config.appDisplayName}-${type}`;
-    data = {
+function mergeIntoDict(dest, src) {
+    let src = src || {};
+    let dest = dest || {};
```
Comment (learykara) [inline src/analytics/actions.es6]: `src` and `dest` are args - can't be redefined here

File: src/analytics/actions.es6 [hunk 8]
```diff
@@ -1,20 +1,57 @@
 const log = require('debug')('ui:analytics');
 const result = require('lodash/object/result');
+const Qs = require('qs');
```
Comment (learykara) [inline src/analytics/actions.es6]: does this need to be added to package.json?

File: src/analytics/actions.es6 [hunk 9]
```diff
@@ -1,20 +1,57 @@
...
+        }
+    }
+
+    return dest;
+}
+
+// This event mirrors the call signature of global.analytics.track
+function track(event, properties, options, fn) {
+    // Argument reshuffling, from original library.
+    if (is.fn(options)) fn = options, options = null;
```
Comment (learykara) [inline src/analytics/actions.es6]: what is `is.fn`? it's throwing an error for me locally

File: src/analytics/actions.es6 [hunk 10]
```diff
@@ -1,20 +1,57 @@
...
+const Qs = require('qs');
 
-function track(action, type, data={}) {
-    action = `${action}: ${global.__env.config.appDisplayName}-${type}`;
-    data = {
+function mergeIntoDict(dest, src) {
+    let src = src || {};
+    let dest = dest || {};
+
+    for (var key in src) {
```
Comment (learykara) [inline src/analytics/actions.es6]: `let key in src`

**Comment by learykara** [comment]
I'm getting a lot of errors trying to use this locally and no data flowing through to segment.  Try to get it up and running locally and then we can get this out

File: src/analytics/TrackedLink.jsx (hunk 2 above)
Comment (rkuykendall) [inline src/analytics/TrackedLink.jsx]: Yeah, renaming it to URL. Would there be a better way? I actually didn't write the code just moved it into this function, but it seems to make sense.

Comment (learykara) [inline src/analytics/TrackedLink.jsx]: no I can't come up with a better way either, this is fine

File: src/analytics/TrackedLink.jsx [hunk 11]
```diff
@@ -35,8 +36,11 @@ class TrackedLink extends React.Component {
 
     _handleClick(event) {
         let data = omit(this.props, 'children', 'className', 'href', 'target');
+        data = {url: this.props.href, ...data};
+        let eventName = `clicked-${global.__env.config.appDisplayName}-${this.props.type}`;
```
Comment (learykara) [inline src/analytics/TrackedLink.jsx]: config should come from props

File: src/analytics/actions.es6 [hunk 12]
```diff
@@ -1,20 +1,58 @@
...
+function track(event, properties, options, fn) {
+    // Argument reshuffling, from original library.
+    if (is.fn(options)) fn = options, options = null;
+    if (is.fn(properties)) fn = properties, options = null, properties = null;
+
+    const __env = global.__env || {};
+
+    let appInfo = {
+        app: result(__env.config, 'app.name', '').toLowerCase(),
+        appDisplayName: result(__env.config, 'app.displayName', '').toLowerCase()
```
Comment (learykara) [inline src/analytics/actions.es6]: config should come from props

File: src/analytics/actions.es6 [hunk 13]
```diff
@@ -1,20 +1,58 @@
...
+
+    global.analytics &&
+        global.analytics.track(event, properties, options, fn);
+}
+
+// This event mirrors the call signature of global.analytics.identify
+function identify(id, traits, options, fn) {
+    // Temporary stub, which can be expanded upon to add data.
+    global.analytics &&
+        global.analytics.track.apply(this, arguments);
```
Comment (learykara) [inline src/analytics/actions.es6]: where does `arguments` come from?

File: src/analytics/actions.es6 [hunk 14]
```diff
@@ -1,20 +1,58 @@
...
+}
 
+// This event mirrors the call signature of global.analytics.page
+function page(category, name, properties, options, fn) {
+    // Temporary stub, which can be expanded upon to add data.
     global.analytics &&
-        global.analytics.track(action, data);
+        global.analytics.page.apply(this, arguments);
```
Comment (learykara) [inline src/analytics/actions.es6]: same question about `arugments`

**Comment by learykara** [comment]
a couple questions, but mostly looks good! :+1:

File: src/analytics/actions.es6 (hunk 14 above)
Comment (rkuykendall) [inline src/analytics/actions.es6]: In JS, arguments is all the arguments passed into a function.

Yeah, it didn't make sense to me either, but it's true. Yay magic variable!

File: src/analytics/actions.es6 (hunk 13 above)
Comment (rkuykendall) [inline src/analytics/actions.es6]: Below

File: src/analytics/actions.es6 (hunk 12 above)
Comment (rkuykendall) [inline src/analytics/actions.es6]: Hm, would you mind elaborating? This is a set of functions, not a react component, so I don't think it has props? These wrappers are designed to let us call things like `analytics.track('clicked-big-button')` and then be able to populate that track event with extra data like the app name and user information, etc.

Comment (learykara) [inline src/analytics/actions.es6]: you're right - ignore me - been a long week :grin:

**Pull request merged**